### DISCUSSION TITLE
How do the limit switches work? (Ramps 1.6, Due, GrblHAL, UGS)

### DISCUSSION THREAD

**Discussion by Racaya** [winder/Universal-G-Code-Sender#1543] [Help] [answered]
Hello all,

Currently I am trying to get a Ramps 1.6 with an Arduino Due to move around some stepper motors. I am using GrblHAL test branch and UGS for communication. However, I got some problems with the limit switches. I will post the same question in the grblHAL git since I have no clue where the problem may be lying. 

Using [the typical switches](https://www.banggood.com/Geekcreit-RAMPS-1_4-Endstop-Switch-For-RepRap-Mendel-3D-Printer-With-70cm-Cable-p-958124.html?cur_warehouse=UK) I seem to be unable to get the controller do something useful.

Whenever I press the button, nothing happens. Only when I release, the controller goes into Alarm state. Same if I invert the limit switches. Same behaviour when I short the signal pin to either - or +.

I tried using the UGS setup wizard. There in the limit switches section, the green fields should turn red when the limit switches are pressed. Unfortunately, that does not happen. The controller goes into Alarm state upon release, but the the green field never changes. Here again it doesn't matter whether I invert the switches or not.
The other limit switches can be shorted between signal and - or + or left open, no difference either. 

![Before..](https://user-images.githubusercontent.com/14227479/108748178-96998e00-753e-11eb-8075-1a529cadd57d.png)

![.. after pressing and releasing the limit switch](https://user-images.githubusercontent.com/14227479/108748230-a4e7aa00-753e-11eb-8357-0f02c50c37e9.png)

I tried looking into the firmware, but looks like my C skills are insufficient to understand the inner workings of grblHAL ;(

So, how do the limit switches work? Am I doing something wrong with the cabling? Do they all need to be in a certain state before setup? 

Thank you in advance!

**Comment by breiler** [comment, marked as answer]
See discussion in: https://github.com/terjeio/grblHAL/discussions/264
